Refactor PersistentVolume admission controller plugin to use volume plugins

jsafrane opened the pull request:
- AdmissionController plugins now get VolumePluginMgr and CloudProvider interface at their disposal
- PersistentVolume admission controller plugin uses it to query volume plugins  for labels to apply on created PersistentVolumes.
- VolumePlugins can optionally implement VolumeLabellerPlugin interface to     read the labels from external cloud.

Fixes https://github.com/kubernetes/kubernetes/issues/20955, https://github.com/kubernetes/kubernetes/issues/21041 and https://github.com/kubernetes/kubernetes/issues/21056

<!-- Reviewable:start -->

---

This change is [<img src="http://reviewable.k8s.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/21386)

<!-- Reviewable:end -->

jsafrane commented:
cc: @justinb as original admission controller author, @saad-ali for GCE and @kubernetes/rh-storage for generic review.

markturansky commented on pkg/volume/aws_ebs/aws_ebs.go:
```diff
@@ -34,6 +35,10 @@ import (
 )
 
+const (
+	awsElasticBlockStoreUnderProvisioningName = "dummy-for-provisioning"
```
How about "placeholder-for-provisioning" instead of "dummy"?

Same with the other 2 volumes.  When an admin queries the PV list, I think "placeholder" is a better word to see than "dummy".

jsafrane replied:
Ok, renamed

markturansky commented:
I think adding plugins to the API server is the right approach. It enables the API server to do a lot more things for volumes.  This PR is admission. We could do validation with plugins, too. All good stuff, I think.

Can you break this into smaller PRs that are easier to review?  It's been requested of me before to break down the big changes into smaller PRs so they're easier to understand and quicker to merge.

We also need to hear from the API machinery team for review.

markturansky commented on pkg/volume/util/plugins/probe.go:
```diff
@@ -0,0 +1,80 @@
+	}
+	allPlugins = append(allPlugins, azure_file.ProbeVolumePlugins()...)
+	allPlugins = append(allPlugins, configmap.ProbeVolumePlugins()...)
+	return allPlugins
```
I understand this function is here for re-use but it was purposefully put in the /cmd binaries so that the binary decides what plugins it has.

I don't know if there is a pattern in place for sharing common code like this across binaries. I'm not sure it will always be the same list everywhere either.

saad-ali replied:
I don't feel strongly either way. I can buy that creating the list in the binary makes it explicit, on the other hand this provides a common/default way to get all plugins, which is what most binaries will want to do--if they want something custom, they can replace this with their own custom logic. I'll leave it up to you guys to decide.

jsafrane replied:
I think that choice of "what plugins binary has" is wrong. It should be "what plugins does Kubernetes installation support" - it should be the same in admission controller, provisioner, recycler and kubelet. I'm open to move it to `cmd/util/volumes.go`, but it would be the first in `cmd/util`.

saad-ali replied:
> I think that choice of "what plugins binary has" is wrong. It should be "what plugins does Kubernetes installation support"

Good point

markturansky commented on pkg/volume/util/plugins/probe.go:
```diff
@@ -0,0 +1,80 @@
+	allPlugins = append(allPlugins, azure_file.ProbeVolumePlugins()...)
+	allPlugins = append(allPlugins, configmap.ProbeVolumePlugins()...)
+	return allPlugins
+}
```
Or instead of the above list, maybe something more like the self-registering Admission plugins do?  

I think I prefer a way to disable plugins in the binary and allowing them to self-register. The default list of plugins is all plugins.  This is the same as self-registering.  They are all "on" by default. The difference is knowing +1 more thing about where to register your plugin. If it's in the same file, that's one fewer detail to know.

Is there a specific benefit of an enabled list v. disable list?

Arguably, this refactoring is also a logically separate PR.  The overall complexity of this large PR can be lessened if we carve off pieces like this.

jsafrane replied:
I agree, volume plugin auto-registration makes sense and it would overall simplify volume plugins. I wonder, was there any reason why it was not done from the beginning?

> The overall complexity of this large PR can be lessened if we carve off pieces like this.
 
Is moving of one function that complex?

markturansky commented on plugin/pkg/admission/persistentvolume/label/admission_test.go:
```diff
@@ -152,3 +132,53 @@ func TestAdmission(t *testing.T) {
+	// this mock supports everything but HostPath
+	return !((spec.PersistentVolume != nil && spec.PersistentVolume.Spec.HostPath != nil) ||
+		(spec.Volume != nil && spec.Volume.HostPath != nil))
+
```
dead line

pmorie commented on pkg/volume/plugins.go:
```diff
@@ -129,6 +129,13 @@ type AttachableVolumePlugin interface {
 
+// VolumeLabellerPlugin is an extended interface of VolumePlugin and is used to label persistentvolumes
+// with metadata from external cloud.
+type VolumeLabellerPlugin interface {
```
nit: VolumeLabeler

jsafrane replied:
Renamed

jsafrane commented:
> break down the big changes into smaller PRs so they're easier to understand and quicker to merge.

I don't think this PR is that big. It adds lot of scaffolding to bring volume plugins into admission controllers, but that's hardly for separate PR. In another project I would split the single patch into several several small patches, sadly that's not how Kubernetes work :-(.

saad-ali commented on pkg/volume/volume.go:
```diff
@@ -137,6 +137,13 @@ type Detacher interface {
+type VolumeLabeler interface {
+	// GetLabels returns labels of the volume. This function SHOULD return error
+	// when the volume does not exist.
+	GetLabels() (map[string]string, error)
```
I really don't like how overloaded the term `Label` has become in the kubernetes ecosystem. But, alas, that ship has sailed.

jsafrane replied:
I am leaving it as it is. Would `GetDefaultLabels` be better?

saad-ali replied:
Neither is great. I'm fine leaving it as is.

saad-ali commented on pkg/volume/plugins.go:
```diff
@@ -129,6 +129,13 @@ type AttachableVolumePlugin interface {
 }
 
+// VolumeLabelerPlugin is an extended interface of VolumePlugin and is used to label persistentvolumes
+// with metadata from external cloud.
```
`...is used to label...` is not quite accurate right? Since it's only used to fetch labels.

jsafrane replied:
Updated to `is used to get default labels for PersistentVolumes from external cloud.`

saad-ali commented on cmd/kubelet/app/plugins.go:
```diff
@@ -27,64 +27,11 @@ import (
 	"k8s.io/kubernetes/pkg/kubelet/network/exec"
 	"k8s.io/kubernetes/pkg/kubelet/network/kubenet"
 	// Volume plugins
```
Remove comment if there are no more volume plugins.

jsafrane replied:
Removed

saad-ali commented on pkg/admission/host.go:
```diff
@@ -0,0 +1,104 @@
+	return host.volumePluginMgr
+}
+
+// VolumeHost
```
This is really ugly. The fact that we have to do this indicates that our interfaces are incorrect. We can refactor this when addressing https://github.com/kubernetes/kubernetes/issues/14217

saad-ali commented on pkg/admission/interfaces.go:
```diff
@@ -57,6 +59,16 @@ type Interface interface {
 
+// AdmissionPluginHost is interface between admission plugins and rest of
+// kube-apiserver.
+type AdmissionPluginHost interface {
```
super nit picky: Is `AdmissionPluginHost` really the right name for this? Think think of anything better though.

saad-ali commented:
LGTM mod nits.

I'll leave it to @markturansky and @justinsb for final LGTM.

jsafrane commented:
@k8s-bot test this issue: #19299

justinsb commented:
Sorry for dropping the ball on my part of this.  This is obviously pretty big - is there a chance of getting this into 1.2 still?  I don't think dynamic volume labeling is working right now with Ubernetes Lite (#22532) - I take it this PR would fix it.

I also created #22602 which is a sort of the minimal fix for tagging, at the cost of repeated code.  (It's still WIP)

jsafrane commented:
I think that #21738 can be enough for Kubernetes 1.2.

derekwaynecarr commented:
@pmorie want to take this?

jsafrane committed:
Refactor PersistentVolume admission controller plugin to use volume plugins.

* AdmissionController plugins now get VolumePluginMgr and CloudProvider
interface at their disposal

* PersistentVolume admission controller plugin uses it to query volume plugins
for labels to apply on created PersistentVolumes.

* VolumePlugins can optionally implement VolumeLabelerPlugin interface to
read the labels from external cloud.

Also, admission for OpenStack Cinder volumes was added.

jsafrane commented:
Rebased

jsafrane commented:
sorry, wrong button, reopening

jsafrane commented:
`federation/cmd/federated-apiserver/app` started using admission plugins and it does not implement `volume.VolumeHost` nor `AdmissionPluginHost`. I think it's time to take a break and wait for better volume host interfaces as designed in #14217 instead of spreading mostly empty `VolumeHost` into another component.

chrislovecnm commented:
Is this for 1.3?

derekwaynecarr commented on pkg/admission/host.go:
```diff
@@ -0,0 +1,104 @@
+var _ AdmissionPluginHost = &realAdmissionPluginHost{}
+var _ volume.VolumeHost = &realAdmissionPluginHost{}
+
+func NewAdmissionPluginHost(cloudProvider cloudprovider.Interface, volumePlugins []volume.VolumePlugin) AdmissionPluginHost {
```
can we structure this differently based on the pattern presented here:
https://github.com/kubernetes/kubernetes/pull/26709

derekwaynecarr commented:
Rather than pass AdmissionPluginHost everywhere, I would prefer a pattern where we let a plugin express that it wants or needs that information.  

My preferred pattern for that is here: https://github.com/kubernetes/kubernetes/pull/26709

So AdmissionPluginInitializer would be able to satisfy a number of Wants interfaces, and if your particular plugin Wanted something like this PR does with AdmissionHost it would implement a Wants interface for that similar to how the referenced PR lets you ask for shared informers.

jsafrane commented:
@derekwaynecarr, Ok, I'll rework it to #26709 when it gets merged.

And no, it does not need to be in 1.3.
